opnieuw/retries.py

```diff
@@ -235,33 +222,31 @@ def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
             )
 
             for retry_action in retry_state:
-                if isinstance(retry_action, DoCall):
-                    try:
-                        return f(*args, **kwargs)
+                try:
+                    return f(*args, **kwargs)
 
-                    except retry_on_exceptions as e:
-                        last_exception = e
-
-                elif isinstance(retry_action, DoWait):
-                    sleep_seconds = random.uniform(
-                        retry_action.min_seconds, retry_action.max_seconds
-                    )
+                except retry_on_exceptions as e:
+                    last_exception = e
 
-                    if sleep_seconds > retry_action.seconds_left:
-                        logger.debug(
-                            "Next attempt would be after retry deadline. No point retrying."
-                        )
-
-                        assert (
-                            last_exception is not None
-                        ), "Exception expected if we have a DoWait retry action!"
-                        raise last_exception
+                sleep_seconds = random.uniform(
+                    retry_action.min_seconds, retry_action.max_seconds
+                )
 
+                if sleep_seconds > retry_action.seconds_left:
                     logger.debug(
-                        f"Sleeping for {sleep_seconds:.3f} seconds after "
-                        f"attempt {retry_action.attempts_so_far}"
+                        "Next attempt would be after retry deadline. No point retrying."
                     )
-                    time.sleep(sleep_seconds)
+
+                    assert (
+                        last_exception is not None
+                    ), "Exception expected if we have a DoWait retry action!"
+                    raise last_exception
+
+                logger.debug(
+                    f"Sleeping for {sleep_seconds:.3f} seconds after "
+                    f"attempt {retry_action.attempts_so_far}"
+                )
+                time.sleep(sleep_seconds)
```

Comment:
One side affect of not having the `DoCall` is that now `WaitLessRetryState` will have extra log statements where it did not have them before. Is that intentional?

Ofcourse this could be easily solved by checking if `retry_action == DoWait(0, 0, 0, 0)` and then continueing, maybe with some aliases.

But atleast you should consider the change here.

Reply:
Did not think of this, but might actually be nice?

If we never need to retry we exit immediately and if not we actually communicate that we are going to retry only without waiting.